controllers/openstackdataplanenodeset_controller.go

```diff
@@ -417,21 +417,39 @@ func (r *OpenStackDataPlaneNodeSetReconciler) Reconcile(ctx context.Context, req
 		return ctrl.Result{}, nil
 	}
 
-	// Generate NodeSet Inventory
-	_, err = deployment.GenerateNodeSetInventory(ctx, helper, instance,
-		allIPSets, dnsData.ServerAddresses, dataplaneAnsibleImageDefaults)
+	deploymentExists, isDeploymentReady, err := checkDeployment(helper, instance)
 	if err != nil {
-		errorMsg := fmt.Sprintf("Unable to generate inventory for %s", instance.Name)
-		util.LogErrorForObject(helper, err, errorMsg, instance)
 		instance.Status.Conditions.MarkFalse(
-			dataplanev1.SetupReadyCondition,
+			condition.DeploymentReadyCondition,
 			condition.ErrorReason,
 			condition.SeverityError,
-			dataplanev1.DataPlaneNodeSetErrorMessage,
-			errorMsg)
+			condition.DeploymentReadyErrorMessage,
+			err.Error())
+		Log.Error(err, "Unable to get deployed OpenStackDataPlaneDeployments.")
 		return ctrl.Result{}, err
 	}
 
+	secretName := fmt.Sprintf("dataplanenodeset-%s", instance.Name)
+	_, _, err = secret.GetSecret(ctx, helper, secretName, instance.Namespace)
+	isInventoryMissing := k8s_errors.IsNotFound(err)
+
+	if isInventoryMissing || isDeploymentReady {
+		// Generate NodeSet Inventory
+		_, err = deployment.GenerateNodeSetInventory(ctx, helper, instance,
+			allIPSets, dnsData.ServerAddresses, dataplaneAnsibleImageDefaults)
+		if err != nil {
+			errorMsg := fmt.Sprintf("Unable to generate inventory for %s", instance.Name)
+			util.LogErrorForObject(helper, err, errorMsg, instance)
+			instance.Status.Conditions.MarkFalse(
+				dataplanev1.SetupReadyCondition,
+				condition.ErrorReason,
+				condition.SeverityError,
+				dataplanev1.DataPlaneNodeSetErrorMessage,
+				errorMsg)
+			return ctrl.Result{}, err
+		}
+	}
+
 	// all setup tasks complete, mark SetupReadyCondition True
 	instance.Status.Conditions.MarkTrue(dataplanev1.SetupReadyCondition, condition.ReadyMessage)
```

Comment:
when this `MarkTrue` is run, it's possible the inventory wasn't updated if `!isDeploymentReady`. That seems wrong. The last deployment could also be failed, and `isDeploymentReady` would be `false` there also, and the inventory wouldn't get regenerated.

what about checking if a deployment is in progress, and if so, then requeue much earlier than this before we reset any conditions or set observedGeneration, etc?